File: 1-two-sum/1-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        solution: List[int] = []
        # create a map of all indexes by their value
        dict:Dict[int, List[int]] = {};
        
        for i, val in enumerate(nums):
            if val not in dict:
                dict[val] = [i]
            else:  
                dict[val].append(i);      
        # loop over each num and see if value exists in map
        for i, val in enumerate(nums):
            diff: int = target - val;
            if diff in dict:
                for idx in dict[diff]:
                    # if exists make sure it's not the same as the current index
                    if idx != i:
                        solution.append(i)
                        solution.append(idx)
                        return solution
        return solution              
        

File: 1-two-sum/1-two-sum.cpp
#include <unordered_map>

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> solution;
        // map all values <value: index>
        unordered_map<int, vector<int>> m;
        for (int i = 0; i < nums.size(); i++) {
            if (m.find(nums[i]) == m.end()) {
                pair<int, vector<int>>p;
                vector<int> temp{i};
                p.first = nums[i];
                p.second = temp;
                m.insert(p);
            } else {
                m[nums[i]].push_back(i);
            }
        }
        for (int i = 0; i < nums.size(); i++) {
            int diff = target - nums[i];
            if (m.find(diff) != m.end()) {
                for (int idx : m[diff]) {
                    if (idx != i) {
                        solution.push_back(i);
                        solution.push_back(idx);
                        return solution;
                    }
                }
            }
        }
        return solution;
    }

};